### DISCUSSION TITLE
Episode 123: An Algorithm Taxonomy

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#14] [Episode]
Episode Link: https://adspthepodcast.com/2023/03/31/Episode-123.html

Feel free to write comments or ask questions 🙂

**Comment by axman6** [comment]
For a Haskell developer, this was a really interesting discussion that I was champing at the bit to jump into. A lot of the algorithms that were mentioned fit naturally into the stream fusion framework that’s used in the Haskell vector and (until recently) text libraries. Probably well covered in the paper [Stream Fusion. From Lists to Streams to Nothing at All](https://www.researchgate.net/publication/221241130_Stream_Fusion_From_Lists_to_Streams_to_Nothing_at_All), but I learnt it mostly from @donsbot’s (of Xmonad and making everything go fast on Haskell fame) blog posts like [this](https://donsbot.com/2009/09/07/stream-fusion-for-haskell-arrays/) and [this](https://donsbot.com/2008/06/04/haskell-as-fast-as-c-working-at-a-high-altitude-for-low-level-performance/).

As for categorising the various types of algorithms, of course Haskellers have spent the time coming up with precise, but also obscure, names for them. They can be found in the [recursion-schemes](https://hackage.haskell.org/package/recursion-schemes) package which gives more precise names to generating structures, reducing, transforming and converting them. I’m no expert on the topic so can’t tell you which is a catamorphism or an anamorphism, but there are plenty of blog posts out there that do dive into it.

**Reply by codereport** [reply]
Thanks for these links! I will definitely read them and might parlay what I find into one of our next episodes!
